### PR TITLE
Metric-based mesh adaptation

### PR DESCRIPTION
**Pull request by jwallwork23**
Still a few things to do before this is ready. For discussion.

### CODE REVIEW

**Comment by jwallwork23** [comment]
Note: currently testing using PETSc `firedrake` branch, rebased on top of [`jwallwork23/metric-tweaks`](https://gitlab.com/jwallwork23/petsc/-/tree/jwallwork23/metric-tweaks) - see [PETSc MR 4611](https://gitlab.com/petsc/petsc/-/merge_requests/4611).

**Comment by jcallaham** [comment]
I just came across this and am definitely interested in this functionality - have been trying to work off #1063.  Please let me know if there's anything I can help out with.

**Comment by jwallwork23** [comment]
> I just came across this and am definitely interested in this functionality - have been trying to work off #1063. Please let me know if there's anything I can help out with.

@jcallaham Thanks for your interest and for offering to help out! The functionality works in 2D and 3D serial, still need to fix the 3D parallel interface. Are you able to install using the configure flags I put in `.github/workflows/build.yml`? I am on holiday at the minute, but if you have any feedback on what you think of the API/functionality/lack thereof then I will take a look when I get back.

**Comment by jcallaham** [comment]
Okay, that checks out with what I was able to get working - everything except 3D parallel.  I was using a slightly different configuration (I think mainly pulling an older ptscotch since it seems like parmmg is not yet compatible with version 7).  I'm actually about to take off for a couple of weeks as well, but I'll try out your config and follow up when I get back.

Do you have any sense of what the issue with parallelization is?  As far as I could tell it looked like the reordering in `dmcommon.to_petsc_local_numbering` might be the culprit, but I'm not sure.  I don't have a good understanding of how exactly the ordering in Firedrake and PETSc differ, but I'd be happy to take a crack at getting that to work.

I like the API and functionality a lot, though I did actually have a couple of minor thoughts:
- It seems like `RiemannianMetric.normalise` assumes that the metric is SPD.  If that's true, it could be helpful to add a call to `enforce_spd` in `normalise`.
- Similarly, I don't think `dm_plex_metric_target_complexity` is applied unless `normalise` is specifically called, which could be a little confusing.  Is it safe to assume that if `target_complexity` is specified then the user intends for `normalise` to be applied?
- I wonder if it might be possible to also have a `RiemannianMetric` constructor that accepts a function and metric parameters, rather than the mesh.  Then SPD/normalise/etc could be applied automatically depending on the parameters, which would just simplify the up-front interface a bit.

**Comment by jwallwork23** [comment]
> Okay, that checks out with what I was able to get working - everything except 3D parallel. I was using a slightly different configuration (I think mainly pulling an older ptscotch since it seems like parmmg is not yet compatible with version 7). I'm actually about to take off for a couple of weeks as well, but I'll try out your config and follow up when I get back.
> 
> Do you have any sense of what the issue with parallelization is? As far as I could tell it looked like the reordering in `dmcommon.to_petsc_local_numbering` might be the culprit, but I'm not sure. I don't have a good understanding of how exactly the ordering in Firedrake and PETSc differ, but I'd be happy to take a crack at getting that to work.
> 
> I like the API and functionality a lot, though I did actually have a couple of minor thoughts:
> 
> * It seems like `RiemannianMetric.normalise` assumes that the metric is SPD.  If that's true, it could be helpful to add a call to `enforce_spd` in `normalise`.
> * Similarly, I don't think `dm_plex_metric_target_complexity` is applied unless `normalise` is specifically called, which could be a little confusing.  Is it safe to assume that if `target_complexity` is specified then the user intends for `normalise` to be applied?
> * I wonder if it might be possible to also have a `RiemannianMetric` constructor that accepts a function and metric parameters, rather than the mesh.  Then SPD/normalise/etc could be applied automatically depending on the parameters, which would just simplify the up-front interface a bit.

Yeah I agree that `to_petsc_local_numbering` probably needs replacing - I just copied it from the previous implementation you pointed to. Would be good to discuss how to do it when we are both available.

Thanks for the suggestions on the API - all good points. I have been wondering whether `RiemannianMetric` should just be a subclass of `Function`... would perhaps make things more straightforward.

**Comment by danshapero** [comment]
I'm really excited about this functionality also -- let me know if you want another beta tester!

The API might feel more familiar to what Firedrake users are already accustomed to if all you needed to do was create a Function that lives in a tensor function space, rather than have to separately create a RiemannianMetric. Is it possible to have `adapt` take in a Function and the metric parameters and create a RiemannianMetric internally? Is the average user likely to actually call any of the methods of RiemannianMetric, or are those mostly things needed inside `adapt`?

This is kind of a pedantic edge case, but it looks like the metric parameters are tied to the metric itself in the current code; does that have to be the case? Is it possible that someone might want to use the same tensor field but different values of p?

**Comment by jcallaham** [comment]
@jwallwork23 I do like the idea of subclassing `Function`, especially considering methods like `rename`, `project`, etc.  Anyway I'm back around now and can dig into the reordering (though no rush if you're still out).  Do you know of any documentation that describes the difference between PETSc and Firedrake ordering in any depth?  Or, the original function was suggested for parallel use [here](https://github.com/firedrakeproject/firedrake/pull/1063#discussion_r115437143)... @wence- any chance you have an idea why this may not still be parallel-safe 5 years on?

I was able to build an image using your config and got 13/16 tests passing in `test_meshadapt.py` (`test_intersection[3]`, `test_no_adapt_parallel`, and `test_adapt_parallel_3d` failed, though I think all three of those are expected?).  

@danshapero I also like the idea of a more functional interface, though I could see it being a little tricky to completely hide `RiemannianMetric` given that you might want to do some metric-specific manipulations like intersection or normalization.  Unless those were all done behind the scenes as directed by the parameters, which I think might also take care of your edge case...

**Comment by jwallwork23** [comment]
@danshapero Thanks for the suggestions, I'll rework the code so that `RiemannianMetric` is a subclass of `Function`. You are right that the metric parameters are associated with the metric at the minute. In fact, they are associated with its plex. That's because we stash the metric parameters on the plex on the PETSc level. Could consider changing that, if desired.

@jcallaham Yes, those tests fail for me, too. The 3D intersection one I haven't got to the bottom of why, though (it also fails if you make it serial). That'd be great if you are up for looking into the reordering business, thanks. I wrote [some documentation](https://firedrakeproject.org/petsc-interface.html#accessing-the-petsc-mesh-representation) about interfacing, but it only covers what is already in the code here and probably doesn't go into the depth you are looking for. Happy to chat about this at some point.

**Comment by danshapero** [comment]
> You are right that the metric parameters are associated with the metric at the minute. In fact, they are associated with its plex. That's because we stash the metric parameters on the plex on the PETSc level. Could consider changing that, if desired.

Gotcha, thanks for explaining! I didn't understand where the metric parameters get stashed -- if they're stored at the PETSc level then my suggestion to dissociate them doesn't make any sense.

File: firedrake/cython/dmcommon.pyx [hunk 1]
```diff
@@ -2726,6 +2726,40 @@ def prune_sf(PETSc.SF sf):
...
+    for p in range(*section.getChart()):
+        CHKERR(PetscSectionGetDof(section.sec, p, &dof))
+        if dof > 0:
+            CHKERR(PetscSectionGetOffset(section.sec, p, &off))
+            if off >= 0:
+                for d in range(dof):
+                    for k in range(dim):
+                        oarray[idx] = varray[dim*off + d*off + k - start]
+                        idx += 1
+    return out
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply firedrake/cython/dmcommon.pyx:2760]: Thanks Koki, I should have some time to look into this tomorrow.

**Comment by jwallwork23** [comment]
@jcallaham @danshapero I changed the API so that `RiemannianMetric` is a subclass of `Function`, hopefully that's a bit neater. Let me know what you think.

File: firedrake/cython/dmcommon.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwallwork23) [reply firedrake/cython/dmcommon.pyx]: Most recent commits use this version.

**Comment by jwallwork23** [comment]
I did some debugging. First, I read the ASCII output files for the PETSc Vec version of the metric and it seemed to be as expected. Then I tried the following:
```
diff --git a/firedrake/meshadapt.py b/firedrake/meshadapt.py
index 93ab3825..15e2f917 100644
--- a/firedrake/meshadapt.py
+++ b/firedrake/meshadapt.py
@@ -350,8 +350,19 @@ class MetricBasedAdaptor(AdaptorBase):
         :return: a new :class:`MeshGeometry`.
         """
         self.metric.enforce_spd(restrict_sizes=True, restrict_anisotropy=True)
-        metric = self.metric._reordered
-        newplex = self.mesh.topology_dm.adaptMetric(metric, "Face Sets", "Cell Sets")
+
+        # TODO: DEBUG
+        comm = self.mesh.comm.rank
+        nv = self.mesh.num_vertices()
+        sizef = self.metric._reordered.getLocalSize()
+        sizep = self.metric._plex.metricCreate().getLocalSize()
+        print(f"{comm}: #vertices {nv}, Firedrake {sizef} ({sizef/nv}), PETSc {sizep} ({sizep/nv})")
+        exit(0)
+        # viewer = PETSc.Viewer().createASCII("test.ascii", "w")
+        # viewer(self.metric._reordered)
+        # exit(0)
+
+        newplex = plex.adaptMetric(self.metric._reordered, "Face Sets", "Cell Sets")
         return fmesh.Mesh(newplex, distribution_parameters={"partition": False})

     @PETSc.Log.EventDecorator("MetricBasedAdaptor.interpolate")
```
Here `metricCreate` just creates a Vec with 9 dofs per vertex (in this 3D case). On 2 processors, I found that
```
1: #vertices 146, Firedrake 1134 (7.767123287671233), PETSc 1314 (9.0)
0: #vertices 146, Firedrake 810 (5.5479452054794525), PETSc 1314 (9.0)
```
So it seems to me that `metricCreate` creates a Vec that does indeed have 9 dofs per vertex, but the reordered metric Vec that I create using `Function.dat.vec_ro` doesn't have the same shape. Is there something that I misunderstood, @ksagiyam?

**Comment by ksagiyam** [comment]
`Function.dat.vec_ro` does not include halo values; it is called "global" vector (as opposed to  "local" vector which includes ghost values) which has no duplication and is simply partitioned according to the ownership. That is why we needed to use `V.dm.getGlobalSection()` instead of `V.dm.getSection()` in `to_petsc_local_numbering`.

**Comment by ksagiyam** [comment]
You might be right about the error  message `unable to diagonalize at least 1 metric`.

I think people usually work with local vectors, and I think it is in general a good practice. `RiemannianMetric.vec` currently returns global vector; if we want it to return the local vector, we need to make sure that all components using this `vec` method expect the local vector.

You can quickly test if the following removes the error :

Follow what is done in https://github.com/OP2/PyOP2/blob/3a1b62fe38a89ddee4e23bebf664a295f5538990/pyop2/types/dat.py#L691 to make the local Vec; here `size[0]` is the "owned" size (< total size with halo), which makes the output "global" vector. We can use the entire `self._data` with correct total size to make the local vector.

**Comment by ksagiyam** [comment]
but if we make `RiemannianMetric.vec` return the local vec, we at least need to redo `to_petsc_local_numbering`.

**Comment by jwallwork23** [comment]
Pushed some commits based on these ideas. Still get the same errors, but no new ones are introduced. Currently just copied these lines, not sure if they need tweaking.
https://github.com/firedrakeproject/firedrake/blob/7bd5ca2b50e2d6adf13da8d64953a779139e545d/firedrake/meshadapt.py#L158-L165

**Comment by jwallwork23** [comment]
> It's a little hard to tell from the [PETSc source](https://gitlab.com/petsc/petsc/-/blob/main/src/dm/impls/plex/adaptors/parmmg/parmmgadapt.c) but I assume ParMMG would expect ghost values

IIRC ParMmg actually wants only the local portion (no halos). It doesn't need them because it adapts based on the local part of the mesh, repartitions and then repeats these two steps. (Default 3 times total, can be modified with `-dm_plex_metric_num_iterations`.)

**Comment by jcallaham** [comment]
Okay that would make sense about ParMMG not needing halos.  Though I also realized that the error I'm getting with the "global" version (no halos) is `## Error: MMG5_scaleMesh: unable to diagonalize at least 1 metric (point X)` where `X` is always one greater than the number of "global" vertices (i.e. no ghosts).  Is that the same error you're seeing?  That's what makes me wonder if ParMMG is actually expecting the ghost vertices.

As far as the dofs/vertex goes, in your debugging code doesn't `nv = mesh.num_vertices()` give the local size (including ghost vertices)?  I get 9 dofs/vertex if I use `sizef = np.prod(self.metric.dat.data_ro_with_halos.shape)` instead of `self.metric.vec.getSize()` (which gives the same as `self.metric._reordered.getSize()`).

If I create a "local" vec for the renumbering, use `V.dm.getSection()`, and modulo indexing `oarray[idx] = varray[dim*off + d*off + k - start % len(varray)]` in `to_petsc_local_numbering` the parallel test will pass.  Here's the diff:
```
+    @futils.cached_property
+    def _local_vec(self):
+        return PETSc.Vec().createWithArray(self.dat._data, comm=self.function.ufl_domain().comm)
 
     @property
     def reordered_vec(self):
-        return dmcommon.to_petsc_local_numbering(self.vec, self.function_space())
+        vec = dmcommon.to_petsc_local_numbering(self._local_vec, self.function_space())
+        return vec
```

I still get ParMMG errors when I try this out on a slightly more complicated example: `## Error: MMG5_coquilface: Wrong boundary tags: Only 1 boundary face found in the shell of the edge` on the second iteration.  So it might not have actually fixed the problem as much as move it somewhere else...

**Comment by jwallwork23** [comment]
Closing and reopening as #2796.

**Pull request closed without merging**